src/libraries/System.Data.Common/src/System/Xml/XmlDataDocument.cs

```diff
@@ -21,13 +18,13 @@ namespace System.Xml
     [Obsolete("XmlDataDocument class will be removed in a future release.")]
     public class XmlDataDocument : XmlDocument
     {
-        private DataSet _dataSet;
+        private DataSet _dataSet = null!;
```

Comment:
Can you use [MemberNotNull] on the Init and AttachDataSet methods instead? You can get rid of this `null!` if you do.

Reply:
The same for `_pointers`, `_columnChangeList`, `_foliationLock` and `_attrXml`.

Reply:
I tried that. The problem is that there is this constructor as well:

```C#
        internal XmlDataDocument(XmlImplementation imp) : base(imp)
        {
        }
```

Reply:
We may not be entirely consistent, but anywhere we've used `= null!;` on fields we've tried to add a comment next to it indicating why.

Reply:
> The problem is that there is this constructor as well

What ensures _dataSet is set to non-null when that ctor is used?  Or the field is never used when that ctor is used?

Reply:
> What ensures _dataSet is set to non-null when that ctor is used? Or the field is never used when that ctor is used?

I believe the field is never used in that instance. It's instantiated by `XmlDataImplementation` which exposes it as `XmlDocument`.